python/pyiceberg/expressions/base.py

```diff
@@ -474,3 +474,77 @@ def visit_unbound_predicate(self, predicate) -> BooleanExpression:
 
     def visit_bound_predicate(self, predicate) -> BooleanExpression:
         raise TypeError(f"Found already bound predicate: {predicate}")
+
+
+class BoundBooleanExpressionVisitor(Generic[R], BooleanExpressionVisitor[R], ABC):
+    def visit_in(self, ref: BoundReference[T], literals: Tuple[Literal[T], ...]) -> R:
+        """Visit the ref and literals from an In boolean expression
+
+        Args:
+            ref (BoundReference[T]): The ref used in the In boolean expression
+            literals (Tuple[Literal[T], ...]): The literals used in the In boolean expression
+
+        Raises:
+            NotImplementedError: If the concrete bound boolean expression visitor does not override this method
+
+        Returns:
+            R: The return type defined by the concrete bound boolean expression visitor
+        """
+        raise NotImplementedError(f"{type(self).__name__} does not support visiting In expressions")
+
+    def handle_non_reference(self, term: BoundTerm) -> R:
+        """Handle a non-reference value in this visitor
+
+        Visitors that specificially require references can use this method to return a default value for expressions with non-references.
+        The default implementation will throw a TypeError because the non-reference is not supported.
+
+        Args:
+            term (BoundTerm): a non-reference term such as a transform
+
+        Raises:
+            TypeError: when handle_non_reference is not implemented in the concrete visitor
+
+        Returns:
+            R: The return type defined by the concrete bound boolean expression visitor
+        """
+        raise TypeError(f"BoundExpressionVisitor does not support non-reference: {term}")
+
+    def visit_unbound_predicate(self, predicate: UnboundPredicate[T]):
+        """Visit an unbound predicate
+
+        Args:
+            predicate (UnboundPredicate[T]): An unbound predicate
+
+        Raises:
+            TypeError: This always raises since an unbound predicate is not expected in a bound boolean expression
+        """
+        raise TypeError(f"Not a bound predicate: {predicate}")
+
+    @singledispatchmethod
```

Comment:
Keep in mind that the `singledispatchmethod` is much slower than just single dispatch: https://bugs.python.org/issue40988

I'm not sure how big these expressions trees can grow. In the case of auto-generated SQL from some BI tool, I think it can become significant.

Reply:
That's right, I forgot about the performance hit when using this. I replaced it with a simple if block that does an `isinstance` check.